### PR TITLE
rec: QType ADDR is supposed to be used internally only.

### LINKED ISSUE #11337: rec: ADDR queries for nameservers, probably from refetch
- Program: Recursor
 - Issue type: Bug report

### Short description
I'm running Recursor with `max-cache-ttl=172800`. I turned on `refresh-on-ttl-perc=10` about 45-46 hours ago.

Recursor recently started to make `A`, `AAAA` and `ADDR` queries for TLD nameservers.

### Environment
 - Operating system: Ubuntu 20.04 (amd64)
 - Software version: 4.7.0~alpha0+master.488.g4becceb8a-1pdns.focal
 - Software source: PowerDNS repository

### Steps to reproduce
1. Configure Recursor as above
2. `tcpdump` DNS traffic
3. `dig horse ns && sleep 47h && dig horse ns`

I think.

### Expected behaviour
No `ADDR` queries.

### Actual behaviour
`ADDR` queries.

### Other information
```
2022-02-18 14:51:32.661669 IP 45.79.214.107.56217 > 192.33.4.12.53: 18880 [1au] NS? horse. (34)
2022-02-18 14:51:32.679549 IP 192.33.4.12.53 > 45.79.214.107.56217: 18880- 0/7/9 (661)
2022-02-18 14:51:32.680749 IP 45.79.214.107.2197 > 37.209.198.10.53: 41209 [1au] NS? horse. (34)
2022-02-18 14:51:32.699218 IP 37.209.198.10.53 > 45.79.214.107.2197: 41209*- 5/0/1 NS c.nic.horse., NS a.nic.horse., NS b.nic.horse., NS d.nic.horse., RRSIG (304)
2022-02-18 14:51:34.201614 IP 45.79.214.107.47235 > 37.209.196.10.53: 49139 [1au] A? nic.horse. (38)
2022-02-18 14:51:34.214129 IP 37.209.196.10.53 > 45.79.214.107.47235: 49139*- 3/0/1 A 52.44.201.210, A 3.232.3.131, RRSIG (271)
2022-02-18 14:51:34.220342 IP6 2600:3c02:e000:74::53c:1.11325 > 2001:dcd:1::10.53: 28843 [1au] Type65400? a.nic.horse. (40)
2022-02-18 14:51:34.231840 IP6 2001:dcd:1::10.53 > 2600:3c02:e000:74::53c:1.11325: 28843*- 0/4/1 (580)
2022-02-18 14:51:35.736299 IP6 2600:3c02:e000:74::53c:1.7862 > 2001:dcd:3::10.53: 14246 [1au] Type65400? b.nic.horse. (40)
2022-02-18 14:51:35.747754 IP6 2001:dcd:3::10.53 > 2600:3c02:e000:74::53c:1.7862: 14246*- 0/4/1 (582)
2022-02-18 14:51:37.250097 IP 45.79.214.107.58266 > 37.209.192.10.53: 10734 [1au] Type65400? c.nic.horse. (40)
2022-02-18 14:51:37.261752 IP 37.209.192.10.53 > 45.79.214.107.58266: 10734*- 0/4/1 (582)
2022-02-18 14:51:40.766247 IP6 2600:3c02:e000:74::53c:1.13109 > 2001:dcd:2::10.53: 20012 [1au] Type65400? d.nic.horse. (40)
2022-02-18 14:51:40.777890 IP6 2001:dcd:2::10.53 > 2600:3c02:e000:74::53c:1.13109: 20012*- 0/4/1 (582)
2022-02-18 14:51:44.282080 IP6 2600:3c02:e000:74::53c:1.13086 > 2001:dcd:4::10.53: 21940 [1au] AAAA? c.nic.horse. (40)
2022-02-18 14:51:44.296922 IP6 2001:dcd:4::10.53 > 2600:3c02:e000:74::53c:1.13086: 21940*- 2/0/1 AAAA 2001:dcd:3::10, RRSIG (269)
```

```
$ sudo rg -vN '^(#|$)' /etc/powerdns/recursor.conf
aggressive-nsec-cache-size=10000
allow-from=100.64.0.0/10, 127.0.0.0/8, ::1/128, 2600:3c02::f03c:91ff:fe70:fd08/128, 2600:3c02::f03c:91ff:febc:a28b/128, 2600:3c02::f03c:92ff:fefd:f09a/128, 2600:3c02:e000:74::/64, 2600:3c02:e000:bc::/64, 2600:3c02:e001:1d00::/56, fd7a:115c:a1e0::/48
allow-trust-anchor-query
any-to-tcp
carbon-ourname=ultimate_mattnordhoff_net
carbon-server=2a02:2770:8::2635:0:1
config-dir=/etc/powerdns
dnssec=validate
dnssec-log-bogus
edns-padding-from=100.64.0.0/10, fd7a:115c:a1e0::/48
edns-padding-mode=always
extended-resolution-errors
forward-zones=82.100.in-addr.arpa=2600:3c02:e000:74::53a:1, 83.100.in-addr.arpa=2600:3c02:e000:74::53a:1, 98.100.in-addr.arpa=2600:3c02:e000:74::53a:1, 122.100.in-addr.arpa=2600:3c02:e000:74::53a:1, d.f.ip6.arpa=2600:3c02:e000:74::53a:1
hint-file=/usr/share/dns/root.hints
include-dir=/etc/powerdns/recursor.d
local-address=100.82.87.107, 127.0.0.1, ::1, 2600:3c02:e000:74::53c:0, fd7a:115c:a1e0:ab12:4843:cd96:6252:576b
log-common-errors
lua-config-file=/etc/powerdns/recursor.lua
max-cache-entries=100000
max-cache-ttl=172800
max-negative-ttl=10800
max-packetcache-entries=50000
non-local-bind
query-local-address=0.0.0.0, ::
quiet=yes
refresh-on-ttl-perc=10
setgid=pdns
setuid=pdns
tcp-fast-open-connect
threads=1
udp-truncation-threshold=1452
```

```
$ cat /etc/powerdns/recursor.lua
dofile("/usr/share/pdns-recursor/lua-config/rootkeys.lua")

addNTA("82.100.in-addr.arpa")
addNTA("98.100.in-addr.arpa")
addNTA("122.100.in-addr.arpa")
addNTA("d.f.ip6.arpa")
```

### PR DESCRIPTION
**Pull request by omoerbeek**
Should fix #11337

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/rec-taskqueue.cc [hunk 1]
```diff
@@ -87,6 +87,14 @@ void runTaskOnce(bool logErrors)
 
 void pushAlmostExpiredTask(const DNSName& qname, uint16_t qtype, time_t deadline)
 {
+  switch (qtype) {
+    // Internal types
+  case QType::ENT:
+  case QType::ADDR:
+  case QType::ALIAS:
+  case QType::LUA:
+    return;
```
Comment (rgacogne) [inline pdns/recursordist/rec-taskqueue.cc:96]: We might even want to log a loud message here?

**Comment by omoerbeek** [comment]
There's an issue with logging: it's possible to trigger externally, repeating 

`dig -p 5301  www.drijf.net -t TYPE65401'

leads to log spam:

`Feb 21 08:59:10 Cannot push task for ALIAS`

So this should be max log level DEBUG I suppose, or have some other safeguard like `!log-common-errors`

**Comment by rgacogne** [comment]
> There's an issue with logging: it's possible to trigger externally, repeating
> 
> `dig -p 5301 www.drijf.net -t TYPE65401`
> 
> leads to log spam:
> 
> `Feb 21 08:59:10 Cannot push task for ALIAS`
> 
> So this should be max log level DEBUG I suppose, or have some other safeguard like `!log-common-errors`

Don't we rather want to filter queries for these types upfront?

**Comment by omoerbeek** [comment]
> > There's an issue with logging: it's possible to trigger externally, repeating
> > `dig -p 5301 www.drijf.net -t TYPE65401`
> > leads to log spam:
> > `Feb 21 08:59:10 Cannot push task for ALIAS`
> > So this should be max log level DEBUG I suppose, or have some other safeguard like `!log-common-errors`
> 
> Don't we rather want to filter queries for these types upfront?

Right, we can make a function using a list of qtypes that we don't want to answer and use that in the incoming path and taskqueue.

**Comment by omoerbeek** [comment]
From the commit message of the last commit:

While working on this, I noticed inconsistencies between UDP and TDP logging:
1 uses g_quiet, the other g_logCommonErrors. This needs to be revisited.

Also: the use of MT->getTid() seems strange, as it is used in places where
no mthread has been started yet indoProcessUDPQuestion() and handleRunningTCPQuestion().
That is, unless I'm missing something.

File: pdns/recursordist/rec-taskqueue.cc [hunk 2]
```diff
@@ -181,13 +181,23 @@ void runTaskOnce(bool logErrors)
 
 void pushAlmostExpiredTask(const DNSName& qname, uint16_t qtype, time_t deadline)
 {
+  if (SyncRes::isUnsupported(qtype)) {
+    auto log = g_slog->withName("taskq")->withValues("name", Logging::Loggable(qname), "qtype", Logging::Loggable(QType(qtype).toString()));
```
Comment (chbruyand) [inline pdns/recursordist/rec-taskqueue.cc:185]: `Loggable<>::to_string()` template could use a `QType` specialization as it is a common type we may use in logging ?

**Review by chbruyand — APPROVED** [review]
LGTM!

**Review by rgacogne — APPROVED** [review]

File: pdns/syncres.hh [hunk 3]
```diff
@@ -736,6 +736,19 @@ public:
...
   }
 
+  static bool isUnsupported(QType qtype)
+  {
+    switch (qtype.getCode()) {
+      // Internal types
+    case QType::ENT:
+    case QType::ADDR:
+    case QType::ALIAS:
+    case QType::LUA:
```
Comment (Habbie) [inline pdns/syncres.hh:746]: Other than "looks weird in logs", there is no reason to treat ALIAS and LUA different in the Recursor. How about not loading those two types at all?

Comment (omoerbeek) [reply pdns/syncres.hh:746]: Well, somebody could still specify a query with that qtype, type being loaded or not.

**Pull request merged**